File: android/app/src/main/java/com/hermexapp/android/model/ServerInfo.kt
package com.hermexapp.android.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// Shapes verified against the pinned upstream source (root AGENTS.md hard rule #1):
// `_handle_health` and the `/api/auth/*` handlers in `api/routes.py`. Every field
// is nullable with a default so unknown or renamed upstream fields never crash
// decoding (hard rule #3); ApiClient decodes with `ignoreUnknownKeys = true`.

/** Subset of `GET /health` the app reads — mirrors the iOS `HealthResponse`. */
@Serializable
data class HealthResponse(
    val status: String? = null,
    val sessions: Int? = null,
    @SerialName("active_streams") val activeStreams: Int? = null,
    @SerialName("uptime_seconds") val uptimeSeconds: Double? = null,
)

/** `GET /api/auth/status` — mirrors the iOS `AuthStatusResponse`. */
@Serializable
data class AuthStatusResponse(
    @SerialName("auth_enabled") val authEnabled: Boolean? = null,
    @SerialName("logged_in") val loggedIn: Boolean? = null,
    // Finer-grained capabilities newer servers report. `password_auth_enabled ==
    // false` (and only an explicit false) marks a passkey-only server the app
    // can't sign into; a missing value means "unknown" → treat as password auth.
    @SerialName("password_auth_enabled") val passwordAuthEnabled: Boolean? = null,
    @SerialName("passkeys_enabled") val passkeysEnabled: Boolean? = null,
    @SerialName("passwordless_enabled") val passwordlessEnabled: Boolean? = null,
)

/** `POST /api/auth/login` / `logout` result — mirrors the iOS `LoginResponse`. */
@Serializable
data class LoginResponse(
    val ok: Boolean? = null,
    val message: String? = null,
    val error: String? = null,
)

File: HermesMobile/Models/ServerInfo.swift
import Foundation

struct HealthResponse: Decodable {
    let status: String?
    let sessions: Int?
    let activeStreams: Int?
    let uptimeSeconds: Double?
}

struct AuthStatusResponse: Decodable {
    let authEnabled: Bool?
    let loggedIn: Bool?
    /// Finer-grained capabilities newer servers report. All optional so older
    /// servers that omit them decode unchanged. `password_auth_enabled == false`
    /// (and only an explicit false) marks a passkey-only server we can't sign
    /// into yet (#255); a missing value means "unknown" → treat as today.
    let passwordAuthEnabled: Bool?
    let passkeysEnabled: Bool?
    let passwordlessEnabled: Bool?
    /// Set when the server offers single sign-on. `is_auth_enabled()` upstream
    /// covers password, OIDC *and* trusted-header, so "auth on, password off"
    /// on its own never meant passkeys (#3).
    let oidcEnabled: Bool?
    /// Set when an identity proxy in front of the server authenticates the
    /// request (Cloudflare Access, Authentik). Present only when that mode is
    /// on, so nil means "not that kind of server".
    let trustedAuthEnabled: Bool?

    /// The server already considers this client signed in. Trusted-header
    /// deployments authenticate at the proxy, so there is no credential for the
    /// app to send and no login step to perform.
    var isAlreadySignedIn: Bool { loggedIn == true }

    init(
        authEnabled: Bool? = nil,
        loggedIn: Bool? = nil,
        passwordAuthEnabled: Bool? = nil,
        passkeysEnabled: Bool? = nil,
        passwordlessEnabled: Bool? = nil,
        oidcEnabled: Bool? = nil,
        trustedAuthEnabled: Bool? = nil
    ) {
        self.authEnabled = authEnabled
        self.loggedIn = loggedIn
        self.passwordAuthEnabled = passwordAuthEnabled
        self.passkeysEnabled = passkeysEnabled
        self.passwordlessEnabled = passwordlessEnabled
        self.oidcEnabled = oidcEnabled
        self.trustedAuthEnabled = trustedAuthEnabled
    }
}

struct LoginResponse: Decodable {
    let ok: Bool?
    let message: String?
    let error: String?
}
